test/System/Angle.test.cpp

```diff
@@ -25,33 +25,33 @@ TEST_CASE("[System] sf::Angle")
-        STATIC_CHECK(sf::degrees(-180).wrapSigned() == sf::degrees(-180));
-        STATIC_CHECK(sf::degrees(360).wrapSigned() == sf::degrees(0));
-        STATIC_CHECK(sf::degrees(-360).wrapSigned() == sf::degrees(0));
-        STATIC_CHECK(sf::degrees(720).wrapSigned() == sf::degrees(0));
-        STATIC_CHECK(sf::degrees(-720).wrapSigned() == sf::degrees(0));
+        CHECK(sf::degrees(0).wrapSigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(1).wrapSigned() == Approx(sf::degrees(1)));
+        CHECK(sf::degrees(-1).wrapSigned() == Approx(sf::degrees(-1)));
+        CHECK(sf::degrees(90).wrapSigned() == Approx(sf::degrees(90)));
+        CHECK(sf::degrees(-90).wrapSigned() == Approx(sf::degrees(-90)));
+        CHECK(sf::degrees(180).wrapSigned() == Approx(sf::degrees(-180)));
+        CHECK(sf::degrees(-180).wrapSigned() == Approx(sf::degrees(-180)));
+        CHECK(sf::degrees(360).wrapSigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(-360).wrapSigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(720).wrapSigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(-720).wrapSigned() == Approx(sf::degrees(0)));
     }
 
     SECTION("wrapUnsigned()")
     {
         STATIC_CHECK(sf::Angle::Zero.wrapUnsigned() == sf::Angle::Zero);
-        STATIC_CHECK(sf::degrees(0).wrapUnsigned() == sf::degrees(0));
-        STATIC_CHECK(sf::degrees(1).wrapUnsigned() == sf::degrees(1));
-        STATIC_CHECK(sf::degrees(-1).wrapUnsigned() == sf::degrees(359));
-        STATIC_CHECK(sf::degrees(90).wrapUnsigned() == sf::degrees(90));
-        STATIC_CHECK(sf::degrees(-90).wrapUnsigned() == sf::degrees(270));
-        STATIC_CHECK(sf::degrees(180).wrapUnsigned() == sf::degrees(180));
-        STATIC_CHECK(sf::degrees(-180).wrapUnsigned() == sf::degrees(180));
-        STATIC_CHECK(sf::degrees(360).wrapUnsigned() == sf::degrees(0));
-        STATIC_CHECK(sf::degrees(-360).wrapUnsigned() == sf::degrees(0));
-        STATIC_CHECK(sf::degrees(720).wrapUnsigned() == sf::degrees(0));
-        STATIC_CHECK(sf::degrees(-720).wrapUnsigned() == sf::degrees(0));
+        CHECK(sf::degrees(0).wrapUnsigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(1).wrapUnsigned() == Approx(sf::degrees(1)));
+        CHECK(sf::degrees(-1).wrapUnsigned() == Approx(sf::degrees(359)));
+        CHECK(sf::degrees(90).wrapUnsigned() == Approx(sf::degrees(90)));
+        CHECK(sf::degrees(-90).wrapUnsigned() == Approx(sf::degrees(270)));
+        CHECK(sf::degrees(180).wrapUnsigned() == Approx(sf::degrees(180)));
+        CHECK(sf::degrees(-180).wrapUnsigned() == Approx(sf::degrees(180)));
+        CHECK(sf::degrees(360).wrapUnsigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(-360).wrapUnsigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(720).wrapUnsigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(-720).wrapUnsigned() == Approx(sf::degrees(0)));
```

Comment:
These changes show how the new API is perhaps less intuitive. Humans have an easier time reasoning about angles in degrees. It's not intuitive to me that 360 degrees wrapped doesn't yield precisely 0 degrees. At least with radians we know that pi cannot be perfectly represented by floating point numbers so it's not confusing why some number close to 2pi doesn't wrap exactly to 0 radians.

Reply:
float additions never guaranteed no rounding errors and you will never be able to guarantee that 360 degrees wrapped is exactly 0 degrees. It just happened to be that way for these tests but the result was technically speaking undefined behavior.

floating point rounding errors are a very basic concept in computer programming and I feel like if we need to protect users from them in arbitrary situations then we are doing it wrong.

Reply:
> but the result was technically speaking undefined behavior.

Its we're going to get into the pedantry of this, it is not undefined. IEEE float math is perfectly well defined.

> you will never be able to guarantee that 360 degrees wrapped is exactly 0 degrees.

Actually yes we can guarantee that `360.f` will wrap to `0.f`. Floats do a perfect job representing integers. It's once you start using values in between integers or try to use really big numbers that the "weird" stuff comes into play.

Reply:
You are indeed correct that floating point arithmetic rounding errors are well defined, i was not aware of this.

Still, the whole concept of relying on float arithmetic to give an exact result is wild to me.

It is not defined in the sf::Angle documentation anywhere that it will give exact results for some arithmetic operations, therefore in my opinion by not using approx comparisons breaks encapsulation by testing implementation details.